### PR TITLE
doc: Add and update DFU related documentation

### PR DESCRIPTION
**Pull request by markushutzler**
Describes step by step procedure to download firmware over UART.

Signed-off-by: Markus Hutzler markus.hutzler@intel.com

### CODE REVIEW

File: doc/dfu.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+the `README.rst <../bootloader/dm/README.rst>`_ located in `bootloader/dm/`.
+
+Currently, DM functionality consists in the following services provided over
+UART:
+
+ * Application programming
+ * Application erase
+ * System information retrieval (bootloader version, application version, etc.)
+
+Transport Specific Guideleines
```
Comment (jeez) [inline doc/dfu.rst]: small typo

File: doc/dfu.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
+Bootloader: Device Management (aka Firmware Update) Guide
```
Comment (jeez) [inline doc/dfu.rst]: Firmware Manager would describe it better, I believe.

File: doc/dfu.rst [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+This guide explains how to use the Device Management (DM) functionality
+provided by the Quark bootloader.
+
+The QDM (Quark Device Management) design and protocol is described in
+the `README.rst <../bootloader/dm/README.rst>`_ located in `bootloader/dm/`.
+
+Currently, DM functionality consists in the following services provided over
+UART:
+
+ * Application programming
```
Comment (jeez) [inline doc/dfu.rst]: then here I would add the (a.k.a. Firmware Update)

**Comment by jeez** [comment]
LGTM, @markushutzler ! Just added some minor comments, ok?

File: doc/dfu/quark_dfu_uart.rst [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+functionality to flash the LED-blink (blinky) example app to a Quark SE
+development board.
+
+Simple way (using 'make flash'):
+ * Reset the device while connecting the DM GPIO to GND.
+ * Compile, upload and run the example app:
+   ::
+
+       cd examples/blinky
+       make flash SERIAL_PATH=[path_to_serial_interface]
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: Can we add the FULL command e.g. 

make flash SERIAL_PORT=[path_to_serial_interface] SOC=[SOC Name] TARGET=[Target Name]

File: doc/dfu/quark_dfu_uart.rst [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ * Having the possibility to connect the DM GPIO pin to GND or to press the DM
+   Button.
+
+The DM GPIO pin is the pin used to put the device into DM mode. The device has
+to be reset while the GPIO is connected to GND. On Quark Developer Boards a
+button connects the DM GPIO to GND if pressed.
+
+Software Setup
+**************
+
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:21]: Do we need to reference the toolchain here -- it's required to build the ROM and the example.

File: doc/dfu/quark_dfu_uart.rst [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+================================
+
+In order to use the Device Management (DM) functionality over UART on the target
+platform, you must install a ROM with such functionality enabled. To do that,
+perform the following steps:
+
+ * Compile the ROM code using the DM_ENABLE flag:
+   ::
+
+       make rom ENABLE_DM=1
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: Can we add separate command for each ROM ?

make rom SOC=quark_se ENABLE_DM=1
make rom SOC=quark_d2000 ENABLE_DM=1

File: doc/dfu/quark_dfu_uart.rst [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+The hardware setup consists of two main steps:
+
+ * Connecting the board to a Linux host using UART.
+ * Having the possibility to connect the DM GPIO pin to GND or to press the DM
+   Button.
+
+The DM GPIO pin is the pin used to put the device into DM mode. The device has
+to be reset while the GPIO is connected to GND. On Quark Developer Boards a
+button connects the DM GPIO to GND if pressed.
+
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: Do we need to mention WHICH GPIO's / Button.

SW2 on Intel® Quark™Developer Kit D2000 or SW1 on Intel® Quark™ SE Microcontroller C1000

AP_GPIO_AON4 for Intel® Quark™ SE Microcontroller C1000 Developer Board(pin J14.43)
F2_DIO_05 for Intel® Quark™ Developer Kit D2000 (pin J4.6)

File: doc/dfu/quark_dfu_uart.rst [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+dfu-util-qda is the tool needed to flash QFU images to the target or manage the
+device.
+
+ * Clone the qm-dfu-util repository
+   ::
+
+        git clone https://github.com/quark-mcu/qm-dfu-util.git
+
+ * Ensure `dh-autoreconf` is installed.
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:50]: python is required as well ? Do we need to mention that ?

File: doc/dfu/quark_dfu_uart.rst [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+Step by step process (without make flash):
+ * Go to the example directory:
+   ::
+
+       cd examples/blinky
+
+ * Build the project:
+   ::
+
+       make
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: Can we expand this to include the SOC/Core being used ?

D2000:
make SOC=quark_d2000
SE C1000 LMT Core:
make SOC=quark_se TARGET=x86
SE C1000 ARC Core:
make SOC=quark_se TARGET=sensor

File: doc/dfu/quark_dfu_uart.rst [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+ * Create the dfu file:
+   ::
+
+       python ../../tools/sysupdate/qm_make_dfu.py -v ./release/quark_se/x86/bin/blinky.bin -p 1
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:91]: Same here -- can we include the command for D2000 and BOTH SE C1000 cores
D2000:
python ../../tools/sysupdate/qm_make_dfu.py -v ./release/quark_d2000/x86/bin/blinky.bin -p 1
SE C1000 LMT :
python ../../tools/sysupdate/qm_make_dfu.py -v ./release/quark_se/x86/bin/blinky.bin -p 1
SE C1000 ARC 👍 
python ../../tools/sysupdate/qm_make_dfu.py -v ./release/quark_se/sensor/bin/blinky.bin -p 2

File: doc/dfu/quark_dfu_uart.rst [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            Blocks:      2
+            DFU CRC:     0x...
+      qm_make_dfu.py: blinky.dfu written
+
+   blinky.dfu is your generated QFU image.
+ * Reset the device while connecting the DM GPIO to GND.
+ * Download the image:
+   ::
+
+       dfu-util-qda -D ./release/quark_se/x86/bin/blinky.bin.dfu -p [path_to_serial_interface] -R -a 1
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:115]: Same here ..
D2000:
dfu-util-qda -D ./release/quark_d2000/x86/bin/blinky.bin.dfu -p [serial_path] -R -a 1
SE C1000 LMT:
dfu-util-qda -D ./release/quark_se/x86/bin/blinky.bin.dfu -p [serial_path] -R -a 1
SE C1000 ARC:
dfu-util-qda -D ./release/quark_se/sensor/bin/blinky.bin.dfu -p [serial_path] -R -a 2

File: doc/dfu/quark_dfu_uart.rst [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+   The -R parameter will reset the device after the download is finished.
```
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:120]: Does it actually reset -- I thought we needed to manually reset the board ?

File: doc/dfu.rst [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+ * `Quark DFU UART Guide <dfu/quark_dfu_uart.rst>`__
+
+Platform Specific Information
+*****************************
+
++------------------+------------+---------------+-------------+
+| Platform         | UART Port  | DM GPIO       | DM Button   |
++==================+============+===============+=============+
+| Quark D2000 DP   | UART 0     | GPIO2 (J4.5)  | SW2         |
```
Comment (mcomyns) [inline doc/dfu.rst]: GPIO 2 is J4.6  on the D2000 Developer Board

File: doc/dfu/quark_dfu_uart.rst (hunk 5 above)
Comment (jeez) [inline doc/dfu/quark_dfu_uart.rst:21]: We can just link to the section that covers this on the main README.rst file, right? "For environment setup, please refer to (...)."

Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:21]: Yes - that's all we need ...

File: doc/dfu.rst (hunk 13 above)
Comment (markushutzler) [inline doc/dfu.rst]: Thanks, I will change this.

File: doc/dfu/quark_dfu_uart.rst (hunk 5 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:21]: The assumption is that this file extends README.rst.

File: doc/dfu/quark_dfu_uart.rst (hunk 6 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst]: Lets just change it to: 
`make rom SOC=[SOC Name] ENABLE_DM=1`
?

File: doc/dfu/quark_dfu_uart.rst (hunk 8 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:50]: Python is a dependency for qmsi.

File: doc/dfu/quark_dfu_uart.rst (hunk 4 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst]: :thumbsup:

File: doc/dfu/quark_dfu_uart.rst (hunk 9 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst]: Lets just change it to: 
`make  SOC=[SOC Name] TARGET=[Target Name]`
?

Having more then one possible lines of commands might be confusing the first time. 
(for a step by step guide)

File: doc/dfu/quark_dfu_uart.rst (hunk 10 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:91]: Can we just add a note:
- Note: The binary-path may differ if built for Quark SE. 
- Note: ARC binaries need to be downloaded to partition 2. (`-p 2`)

File: doc/dfu/quark_dfu_uart.rst (hunk 11 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:115]: Can we just add a note:
- Note: The binary-path may differ if built for Quark SE .
- Note: ARC binaries need to be  downloaded to alternate setting 2. (`-a 2`)

File: doc/dfu/quark_dfu_uart.rst (hunk 12 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:120]: Yes it does.

File: doc/dfu/quark_dfu_uart.rst (hunk 6 above)
Comment (jeez) [inline doc/dfu/quark_dfu_uart.rst]: +1 for that

File: doc/dfu/quark_dfu_uart.rst (hunk 9 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: yes - that's fine

File: doc/dfu/quark_dfu_uart.rst (hunk 10 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:91]: sure - that will do ... Was just concerned about the ARC one really

Comment (jeez) [inline doc/dfu/quark_dfu_uart.rst:91]: looks cleaner, indeed...

File: doc/dfu/quark_dfu_uart.rst (hunk 11 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:115]: sure - that's ok . Again the main concern I had was we were missing the partition for the ARC

File: doc/dfu/quark_dfu_uart.rst (hunk 12 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:120]: Ok - Previous instruction we got from Daniele on July 7th was 

The user now has to reset the board manually (e.g., by pressing the reset button)

File: doc/dfu/quark_dfu_uart.rst (hunk 6 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst]: yep - that's ok

File: doc/dfu/quark_dfu_uart.rst (hunk 12 above)
Comment (markushutzler) [inline doc/dfu/quark_dfu_uart.rst:120]: Indeed, the user has to reset the board manually to enter DFU mode. To reset in DFU mode `-R` can be used.

File: doc/dfu/quark_dfu_uart.rst (hunk 10 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:91]: Can you change that first note -- the path in the example is for Quark SE

•Note: The binary-path may differ if built for D2000.

File: doc/dfu/quark_dfu_uart.rst (hunk 11 above)
Comment (mcomyns) [inline doc/dfu/quark_dfu_uart.rst:115]: Can you change that first note -- the path in the example is for Quark SE

•Note: The binary-path may differ if built for D2000. 

Once that change is done then it's good to go

**Comment by markushutzler** [comment]
Note: 
- Changed Quark SE -> Quark D2000
- Changed extra note (Line 121) (DM Pin -> DM GPIO)

**Comment by jeez** [comment]
Is this ready for being merged, @markushutzler and @mcomyns ?

**Comment by mcomyns** [comment]
Looks OK for me

**Pull request merged**